### ISSUE TITLE
Add a Media Query event listener

### ISSUE THREAD

**Issue by EdCharbeneau** [EdCharbeneau/BlazorSize#11]
Decide to how to implement media query event listener.

Research:
https://medium.com/better-programming/how-to-use-media-queries-programmatically-in-react-4d6562c3bc97

**Comment by EdCharbeneau** [comment]
Added experiment https://github.com/EdCharbeneau/BlazorSize/tree/experiment-mql

**Comment by EdCharbeneau** [comment]
API design so far:

To properly create a media query and then identify it by its media value, the media value needs to be set by the browser. 

![mediaquery](https://user-images.githubusercontent.com/945168/78945079-6e6f8080-7a8d-11ea-8849-91e6eb512efd.jpg)

This is difficult because JavaScript isn't available until the component has finished rendering. Because of this, construction requires two steps.

https://github.com/EdCharbeneau/BlazorSize/blob/91f12f52f910b2144e2b49e844d2aeb89f7e2142/BlazorSize.CsbExample/Pages/FetchData.razor#L28

https://github.com/EdCharbeneau/BlazorSize/blob/91f12f52f910b2144e2b49e844d2aeb89f7e2142/BlazorSize.CsbExample/Pages/FetchData.razor#L43

A pub/sub pattern is the choice for tracking if resources can be disposed of.
https://github.com/EdCharbeneau/BlazorSize/blob/experiment-mql/BlazorSize/MediaQuery/MediaQueryListener.cs

This is done to ensure that event handlers aren't left in JS after a component is cleaned up.

Finally, a single event notifies the component if any MediaQueries in the list are triggered.
https://github.com/EdCharbeneau/BlazorSize/blob/91f12f52f910b2144e2b49e844d2aeb89f7e2142/BlazorSize.CsbExample/Pages/FetchData.razor#L50

The user is required to handle the event and arguments by manually updating the status of the media query that was returned by the event handler.

**Comment by EdCharbeneau** [comment]
Removed the idea of the object being an inject-able service. Instead, the object should be created per component so a DotnetReference can be tracked and later used for disposal.

https://github.com/EdCharbeneau/BlazorSize/blob/experiment2-mql/BlazorSize.CsbExample/Pages/FetchData.razor

**Comment by jspuij** [comment]
Doesn't that make the "one js interop per mediaquery impossible"? E.g. when 5 components subscribe to (min 768px), you get 5 interop calls. Don't know if this is a design requirement for you, but for me it was.

**Comment by EdCharbeneau** [comment]
You should be able to accomplish both using this methodology. The responsibility of this will be on the developer.

**Comment by EdCharbeneau** [comment]
What is likely possible is a wrapper that can be placed at the root level. Children can then subscribe to it using a CascadingValue.

```
<MediaQuery values>
@body
</MediaQuery>

**Comment by jspuij** [comment]
That’s true. Cascading Values is what I did in my app, which indeed makes the service / delegate stuff redundant.

**Comment by EdCharbeneau** [comment]
A much better API has developed with this new approach.

Here is the new way to declare breakpoints.

MainLayout.razor
```razor
@inherits LayoutComponentBase
<MediaQueryList>
    <div class="sidebar">
        <NavMenu />
    </div>

    <div class="main">
        <div class="top-row px-4">
            <a href="http://blazor.net" target="_blank" class="ml-md-auto">About</a>
        </div>

        <div class="content px-4">
            @Body
        </div>
    </div>
</MediaQueryList>

```
SampleComponent.razor
```razor
<ul class="list-group">
    <li class="list-group-item @Active(IsLarge)">Large</li>
    <li class="list-group-item @Active(IsMedium)">Medium</li>
    <li class="list-group-item @Active(IsSmall)">Small</li>
</ul>

<MediaQuery Media="@Breakpoints.LargeUp" @bind-Matches="IsLarge" />
<MediaQuery Media="@Breakpoints.OnlyMedium" @bind-Matches="IsMedium" />
<MediaQuery Media="@Breakpoints.SmallDown" @bind-Matches="IsSmall" />

@code {
    Func<bool, string> Active = state => state ? "active" : "";
    bool IsLarge = false;
    bool IsMedium = false;
    bool IsSmall = false;
}
```

**Comment by EdCharbeneau** [comment]
Fixed in BlazorSize 2.0
